net/graphsync_fetcher.go

```diff
@@ -30,6 +30,14 @@ const (
 	requestTimeout = 60 * time.Second
 )
 
+// Fetcher defines an interface that may be used to fetch data from the network.
+type Fetcher interface {
+	// FetchTipSets will only fetch TipSets that evaluate to `false` when passed to `done`,
+	// this includes the provided `ts`. The TipSet that evaluates to true when
+	// passed to `done` will be in the returned slice. The returns slice of TipSets is in Traversal order.
+	FetchTipSets(context.Context, types.TipSetKey, peer.ID, func(types.TipSet) (bool, error)) ([]types.TipSet, error)
```

Comment:
Thinking out loud here:

I believe the only reason we need `peer.ID` as a parameter to this method is because we are not persisting blocks from pubsub (instead we are refetching them). 
If nodes persist the blocks they received over pubsub then I think we can guarantee peers we share a connection with (i.e. that are in the peer tracker) will always have the blocks we are fetching (else how would have we gotten the pubsub message). I imagine the concept of an `initialPeer` can go away once #2962 lands since the peers in the tracker will have the block.
Although I don't think that covers the case of fetching blocks our peers have deemed semantically invalid and thus not stored..

Reply:
> I believe the only reason we need peer.ID as a parameter to this method is because we are not persisting blocks from pubsub

I think this is a little less guaranteed then what you are claiming.  There is a race between (1)  the other peer sending us the pubsub message and us asking for data from that peer and (2) the other peer forwarding the pubsub message and then requesting and receiving the data from one of its peers.

However the way that we use `initialPeer` here is really only to check if this node was the one that mined the block so that we can send a request to ourselves over graphsync in this (and only this case).  The reason for doing this is unfortunately complicated -- requesting for data from ourselves *when we have it* does not block on graph sync, but requesting for data from ourselves when we don't have it *does block* so we can't simply include our peer id in the tracker.

Reply:
In the event that our node is the miner of the block we should avoid going to graphsync all together since we have the block (and its parents) locally. Given the wiring of the syncing code this isn't an option. I am thinking that it should be, and that #2962 moves us in that direction. Do you have similar thoughts on this?

Reply:
100% agreed.  This made more sense with the bitswap fetcher and now is just causing more friction.  I also agree that #2962 is a great place to do this.  Once we have code to handle complete blocks arriving over pubsub that short circuits fetching then we can quickly use it for the case of blocks arriving locally from the mining subsystem.